src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -434,33 +434,33 @@ private DeclarationModifiers MakeModifiers(TypeKind typeKind, BindingDiagnosticB
                 }
             }
 
-            if (this.Name == SyntaxFacts.GetText(SyntaxKind.RecordKeyword))
-            {
-                foreach (var syntaxRef in SyntaxReferences)
-                {
-                    SyntaxToken? identifier = syntaxRef.GetSyntax() switch
-                    {
-                        BaseTypeDeclarationSyntax typeDecl => typeDecl.Identifier,
-                        DelegateDeclarationSyntax delegateDecl => delegateDecl.Identifier,
-                        _ => null
-                    };
-
-                    ReportTypeNamedRecord(identifier?.Text, this.DeclaringCompilation, diagnostics.DiagnosticBag, identifier?.GetLocation() ?? Location.None);
-                }
-            }
-
             return result;
         }
 
-        internal static void ReportTypeNamedRecord(string? name, CSharpCompilation compilation, DiagnosticBag? diagnostics, Location location)
+        internal static bool IsReservedTypeName(string? name)
+        {
+            return name is { Length: > 0 } && name.All(c => char.IsLower(c) && c < 128);
+        }
+
+        internal static void ReportReservedTypeName(string? name, CSharpCompilation compilation, DiagnosticBag? diagnostics, Location location)
         {
-            if (diagnostics is object && name == SyntaxFacts.GetText(SyntaxKind.RecordKeyword) &&
-                compilation.LanguageVersion >= MessageID.IDS_FeatureRecords.RequiredVersion())
+            if (diagnostics is null)
             {
-                diagnostics.Add(ErrorCode.WRN_RecordNamedDisallowed, location, name);
+                return;
             }
-        }
 
+            if (name == SyntaxFacts.GetText(SyntaxKind.RecordKeyword))
+            {
+                if (compilation.LanguageVersion >= MessageID.IDS_FeatureRecords.RequiredVersion())
+                {
+                    diagnostics.Add(ErrorCode.WRN_RecordNamedDisallowed, location, name);
+                }
+            }
+            else if (IsReservedTypeName(name))
+            {
+                diagnostics.Add(ErrorCode.WRN_LowerCaseTypeName, location, name);
+            }
```

Comment:
For a type called `foo`, this will generate:

> The type name 'foo' only contains lower-cased ascii characters. Such names may become reserved for the language.

Let's say in a future version we introduced a keyword called `foo`, this will still generate the same message.

My concern is about diagnostic clarity. The message suggests that `foo` *may become* a keyword (which I interpret as "in future"). But this message can be produced if it's *already* a keyword.

Reply:
Actual [keywords](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/) will not produce this message because they will not parse properly. You'll get `error CS1001: Identifier expected` and not an identifier.

For pseudo-keywords, see logic above for "record". We'll have to adjust this similarly when other words become pseudo-keywords.